### PR TITLE
Lmax pipeliner

### PR DESCRIPTION
**Pull request by volkodavs**
Hey guys, i have added a skeleton for lmax disruptor, this version is not complete, but i prefer to keep PR small :D

### CODE REVIEW

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/Client.java [hunk 1]
```diff
@@ -1,9 +1,7 @@
 package com.matchbook.sdk.core;
 
-import com.matchbook.sdk.core.services.AuthService;
-
 public interface Client extends AutoCloseable {
 
-    AuthService getAuthService();
-
+    @Override
+    void close();
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/Client.java]: What's the point in overriding this method in the interface?

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/Balance.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public Collection<Balance> getContent() {
+        return Collections.singleton(this);
+    }
+
+    @Override
+    public String toString() {
+        return "Balance{" +
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/Balance.java]: `Balance.class.getSimpleName()`

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/BalanceRequest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package com.matchbook.sdk.core.clients.rest.dtos.user;
+
+import java.util.Collections;
+import java.util.Map;
+
+import com.matchbook.sdk.core.clients.rest.dtos.PageableRequest;
+import com.matchbook.sdk.core.clients.rest.dtos.PageableRequestBuilder;
+
+public class BalanceRequest extends PageableRequest {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/BalanceRequest.java]: I don't think this is a pageable request.

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/AccountRequest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.matchbook.sdk.core.clients.rest.dtos.user;
+
+import java.util.Collections;
+import java.util.Map;
+
+import com.matchbook.sdk.core.clients.rest.dtos.PageableRequest;
+import com.matchbook.sdk.core.clients.rest.dtos.PageableRequestBuilder;
+
+public class AccountRequest extends PageableRequest {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/AccountRequest.java]: It's not a pageable request.

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/AccountRequest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    public String resourcePath() {
+        return "account";
+    }
+
+    @Override
+    public Map<String, String> parameters() {
+        return Collections.emptyMap();
+    }
+
+    public static class Builder extends PageableRequestBuilder {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/AccountRequest.java]: Not a pageable builder.

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/BalanceRequest.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    public String resourcePath() {
+        return "account/balance";
+    }
+
+    @Override
+    public Map<String, String> parameters() {
+        return Collections.emptyMap();
+    }
+
+    public static class Builder extends PageableRequestBuilder {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/clients/rest/dtos/user/BalanceRequest.java]: If not a pageable request the builder should not be a pageable builder either.

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/UserDisruptorPipeliner.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    public UserDisruptorPipeliner(Disruptor<UserMessage> disruptor) {
+        this.disruptor = disruptor;
+
+        this.userTransformerHandler = new UserTransformerHandler<>();
+        this.userPublisherHandler = new UserPublisherHandler<>();
+
+        init();
+    }
+
+    private void init() {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/UserDisruptorPipeliner.java]: Maybe `buildDisruptor()` could be a better name.

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/handlers/UserPublisherHandler.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.lmax.disruptor.EventHandler;
+import com.lmax.disruptor.LifecycleAware;
+import com.matchbook.sdk.core.StreamObserver;
+import com.matchbook.sdk.core.disruptor.messages.UserMessage;
+import com.matchbook.sdk.core.model.dataobjects.user.Balance;
+
+public class UserPublisherHandler<T extends UserMessage> implements EventHandler<T>, LifecycleAware {
+
+    private String oldThreadName;
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/handlers/UserPublisherHandler.java]: Why do we need to keep track of the thread name and restore it when the handler stops?

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/messages/UserMessage.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package com.matchbook.sdk.core.disruptor.messages;
+
+import com.matchbook.sdk.core.model.dataobjects.user.Balance;
+
+public class UserMessage extends AbstractDisruptorMessage {
+
+    private static final long serialVersionUID = -8425010391202208013L;
+
+    private Balance balance;
+    private com.matchbook.sdk.core.clients.rest.dtos.user.Balance balanceDTO;
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/messages/UserMessage.java]: Should we rename DTO's using `DTO` suffix?

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/model/dataobjects/events/Sport.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return "Sport{" +
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/model/dataobjects/events/Sport.java]: All the `toString()` must be reviewed :grin:

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorker.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package com.matchbook.sdk.core.workers;
+
+import com.matchbook.sdk.core.StreamObserver;
+import com.matchbook.sdk.core.model.dataobjects.user.Balance;
+
+public interface PullerWorker extends AutoCloseable {
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorker.java]: I think I need an explanation on this one :slightly_smiling_face:

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorkerRest.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+public class PullerWorkerRest implements PullerWorker {
+
+    private final ConnectionManager connectionManager;
+
+    private final ScheduledExecutorService authScheduler;
+    private final ScheduledExecutorService balanceScheduler;
+    private final UserDisruptorPipeliner userDisruptorPipeliner;
+
+    private Disruptor<UserMessage> accountDisruptor;
+
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorkerRest.java]: Extra white space

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorkerRest.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    private void keepAliveSession() {
+        UserRestClient userRestClient = new UserRestClientImpl(connectionManager);
+        LoginRequest loginRequest =
+                new LoginRequest.Builder(connectionManager.getClientConfig().getUsername(),
+                        connectionManager.getClientConfig().getPassword()).build();
+
+        authScheduler.scheduleAtFixedRate(new AuthenticationScheduler(userRestClient, loginRequest),
+                0, 4, TimeUnit.HOURS);
+    }
+
```
Comment (alessandrodalbello) [inline matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/workers/PullerWorkerRest.java]: Many extra lines...

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/UserDisruptorPipeliner.java (hunk 7 above)
Comment (volkodavs) [reply matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/UserDisruptorPipeliner.java]: i will change it to `initDisruptor`

File: matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/messages/UserMessage.java (hunk 9 above)
Comment (volkodavs) [reply matchbook-sdk-core/src/main/java/com/matchbook/sdk/core/disruptor/messages/UserMessage.java]: i don't think so,  i have an idea to make REST API a separate library

**Pull request merged**